packages/client/src/client/Web/serviceWorkerRegistration.ts

```diff
@@ -53,51 +56,78 @@ export function register(config?: Config) {
         })
       } else {
         // Is not localhost. Just register service worker
-        registerValidSW(swUrl, config)
+        registerServiceWorker(swUrl, config)
       }
     })
   }
 }
 
-function registerValidSW(swUrl: string, config?: Config) {
+function addStateChangeHandler(
+  serviceWorkerInstance: ServiceWorker,
+  registration: ServiceWorkerRegistration,
+  registrationConfig: Config | undefined,
+) {
+  serviceWorkerInstance.onstatechange = () => {
+    console.debug(
+      "'installing' SW changed state to: ",
+      serviceWorkerInstance.state,
+    )
+    if (serviceWorkerInstance.state === "installed") {
+      if (navigator.serviceWorker.controller) {
+        // At this point, the updated precached content has been fetched,
+        // but the previous service worker will still serve the older
+        // content until all client tabs are closed.
+        console.debug(
+          "New content is available and will be used when all " +
+            "tabs for this page are closed. See https://cra.link/PWA.",
+        )
+        serviceWorkerInstance.postMessage({ type: "SKIP_WAITING" })
+      } else {
+        // At this point, everything has been precached.
+        // It's the perfect time to display a
+        // "Content is cached for offline use." message.
+        console.debug("Content is cached for offline use.")
+
+        // Execute callback
+        if (registrationConfig && registrationConfig.onSuccess) {
+          registrationConfig.onSuccess(registration)
+        }
+      }
+    } else if (serviceWorkerInstance.state === "activated") {
+      if (registrationConfig && registrationConfig.onUpdate) {
+        registrationConfig.onUpdate(registration)
+      }
+    }
+  }
+}
+
+function registerServiceWorker(swUrl: string, config?: Config) {
   navigator.serviceWorker
     .register(swUrl)
     .then((registration) => {
-      registration.onupdatefound = () => {
+      console.debug("SW Reg: handling ..", registration)
+
+      if (registration.waiting) {
```

Comment:
if we refreshed early or there is a pending SW from an earlier failed activation, then we will have a SW that is stuck in waiting state (installed but unable to proceed until all tabs are closed .. a refresh does not count, per SW guides)